walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
+import java.util.List;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Cases
+    // Options for no attendees - should be available all day.
+    if (request.getAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Duration longer than a day will return empty array.
+    if (request.getDuration() > TimeRange.getTimeInMinutes(23, 59)) {
+      return Arrays.asList();
+    }
+    // Check if there are availabilities for all attendees
+    List<String> attendees = new ArrayList<>();
+    attendees.addAll(request.getAttendees());
+    attendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> allAttendeesAvailabilities = findAvailability(events, request, attendees);
+
+    // If there are no available timeslots for all attendees including optional attendees,
+    // Find availabilities for the mandatory attendees.
+    if (allAttendeesAvailabilities.isEmpty()) {
+      return findAvailability(events, request, request.getAttendees());
+    }
+
+    return allAttendeesAvailabilities;
+  }
+
+  /** Private helper method to find all availabilities of people from the request. */
+  private Collection<TimeRange> findAvailability(
+      Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Find all the unavailable times of the attendees.
+    List<TimeRange> unavailabilities = new ArrayList<>();
+    for (Event event : events) {
+      if (!Collections.disjoint(attendees, event.getAttendees())) {
+        unavailabilities.add(event.getWhen());
+      }
+    }
+    // If there are no unavailable times, return whole day.
+    if (unavailabilities.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    List<TimeRange> mergedUnavailabilities = mergeAllUnavailabilities(unavailabilities);
+    Collections.reverse(mergedUnavailabilities);
+    // Find all available times by checking if gaps between busy time ranges are greater than the
+    // duration.
+    List<TimeRange> availabilities = new ArrayList<>();
+    int start = TimeRange.START_OF_DAY;
+    long duration = request.getDuration();
+    for (TimeRange interval : mergedUnavailabilities) {
+      if (interval.start() - start >= duration) {
+        availabilities.add(TimeRange.fromStartEnd(start, interval.start(), false));
+      }
+      start = interval.end();
+    }
+    // Check if there is a gap between there and the end of the day.
+    if (TimeRange.END_OF_DAY - start >= duration) {
+      availabilities.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+    return availabilities;
+  }
+
+  /**
+   * Private helper method using ArrayDeque to merge overlapping time ranges from an ArrayList.
+   * Merging code from https://www.geeksforgeeks.org/merging-intervals/
+   */
```

Comment:
Can you document the expected result? (@return)

Reply:
Done